### PR TITLE
glsl intrinsic functional test

### PR DESCRIPTION
**Pull request by tgrimesnv**
This adds a test for GLSL intrinsic for functional testing. All combinations of types are passed in and checked for correct return values.

There are several bugs found that need issues created. This code is commented out currently.

For issue: https://github.com/shader-slang/slang/issues/3494

### CODE REVIEW

**Comment by jkwak-work** [comment]
Please mention the issue number on the commit message as [described on our contribution guide](https://github.com/jkwak-work/slang/blob/master/CONTRIBUTION.md#commit-to-the-branch).

File: tests/glsl-intrinsic/intrinsic-functional.slang [hunk 1]
```diff
@@ -0,0 +1,2319 @@
...
+//TEST(compute, vulkan):COMPARE_COMPUTE_EX(filecheck-buffer=VK):-allow-glsl -compute -vk -emit-spirv-directly -output-using-type -profile sm_6_0 -render-feature hardware-device
+
+#define TRIG_FUNCTIONS 0
+#define EXP_FUNCTIONS 1
+#define COMMON_FUNCTIONS 2
+#define PACK_FUNCTIONS 3
+#define GEOMETRIC_FUNCTIONS 4
+#define MATRIX_FUNCTIONS 5
+#define VEC_FUNCTIONS 6
+#define INT_FUNCTIONS 7
```
Comment (jkwak-work) [inline tests/glsl-intrinsic/intrinsic-functional.slang:12]: I think these could be better with enum

File: tests/glsl-intrinsic/intrinsic-functional.slang [hunk 2]
```diff
@@ -0,0 +1,2319 @@
...
+
+
+float AngleAndTrig()
+{
+    float val = 0.0f;
+
+	if( radians(90.0f) > 0.0f )
```
Comment (jkwak-work) [inline tests/glsl-intrinsic/intrinsic-functional.slang:19]: We use four space characters for indentation; not tab characters.

File: tests/glsl-intrinsic/intrinsic-functional.slang [hunk 3]
```diff
@@ -0,0 +1,2319 @@
...
+#define EXP_FUNCTIONS 1
+#define COMMON_FUNCTIONS 2
+#define PACK_FUNCTIONS 3
+#define GEOMETRIC_FUNCTIONS 4
+#define MATRIX_FUNCTIONS 5
+#define VEC_FUNCTIONS 6
+#define INT_FUNCTIONS 7
+
+
+float AngleAndTrig()
```
Comment (jkwak-work) [inline tests/glsl-intrinsic/intrinsic-functional.slang:15]: As described on the GLSL spec document, we should test all of types described as "genFType"
https://registry.khronos.org/OpenGL/specs/gl/GLSLangSpec.4.60.pdf

Please check "tests/glsl-intrinsic/intrinsic-basic.slang" for how different types are tested without copy and paste .
```
__generic<let N : int>
bool dimNTypeFuncs()
{
    typealias genFType = vector<float, N>;
    typealias genDType = vector<double, N>;
    typealias genIType = vector<int, N>;
    typealias genUType = vector<uint, N>;
    typealias genBType = vector<bool, N>;
    typealias vec      = vector<float, N>;
    typealias ivec     = vector<int, N>;
    typealias uvec     = vector<uint, N>;
    typealias bvec     = vector<bool, N>;

    genFType outGenFType;
    genDType outGenDType;
    genIType outGenIType, outGenIType2;
    genUType outGenUType, outGenUType2;
```

File: tests/glsl-intrinsic/intrinsic-functional.slang (hunk 1 above)
Comment (tgrimesnv) [reply tests/glsl-intrinsic/intrinsic-functional.slang:12]: The problem with enum with these is in the main function I'd have to do this:
outputBuffer[(uint)IntrinsicType::Trigonometry] = AngleAndTrig();

Which is a lot less clean than:
outputBuffer[EXP_FUNCTIONS] = Exponential();

File: tests/glsl-intrinsic/intrinsic-functional.slang (hunk 2 above)
Comment (tgrimesnv) [reply tests/glsl-intrinsic/intrinsic-functional.slang:19]: Looks like my editor wasn't set to insert spaces for tabs like I usually have it. I'll fix that.

File: tests/glsl-intrinsic/intrinsic-functional.slang (hunk 1 above)
Comment (jkwak-work) [reply tests/glsl-intrinsic/intrinsic-functional.slang:12]: Can you use `[UnscopedEnum]` that is [described in our doc](https://github.com/shader-slang/slang/blob/4b3f554a58e4224806c31d66874fbe60f1f09332/docs/user-guide/02-conventional-features.md)?

**Comment by csyonghe** [comment]
Can you locate the problematic functions that caused the failure in direct to spirv backend, and create issues for them?

**Comment by tgrimesnv** [comment]
> Can you locate the problematic functions that caused the failure in direct to spirv backend, and create issues for them?

Yeah I'm planning on doing that.

**Comment by jkwak-work** [comment]
I think we should close this.
The testing code had been changed over time too much that this PR is not going yo merge easily.

I think we should find CTS test for the functional tests and enable them for our nightly CTS CI.

**Pull request closed without merging**